algebird-core/src/main/scala/com/twitter/algebird/QTree.scala

```diff
@@ -108,32 +152,45 @@ class QTreeSemigroup[A](k: Int)(implicit val underlyingMonoid: Monoid[A]) extend
   }
 }
 
-case class QTree[A](
+case class QTree[@specialized(Int, Long, Float, Double, Unit) A] private[algebird] (
```

Comment:
what about the unapply method on the object? We could lose the sugar of case class, but we still get source compatibility by adding manual apply/unapply/toString/equals/hashCode